astropy/utils/data.py

```diff
@@ -977,145 +1059,86 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
+            errors[source_url] = e
+        except socket.timeout as e:
+            # this isn't supposed to happen, but occasionally a socket.timeout
+            # gets through.  It's supposed to be caught in `urrlib2` and raised
+            # in this way, but for some reason in mysterious circumstances it
+            # doesn't. So we'll just re-raise it here instead
+            errors[source_url] = e
+    else:   # No success
+        raise urllib.error.URLError(
+            "Unable to open any source! Exceptions were {}".format(errors)) \
+            from errors[sources[0]]
 
-        if cache:
-            _acquire_download_cache_lock()
-            try:
-                with shelve.open(urlmapfn) as url2hash:
-                    # We check now to see if another process has
-                    # inadvertently written the file underneath us
-                    # already
-                    if url_key in url2hash:
-                        return url2hash[url_key]
-                    local_path = os.path.join(dldir, hash.hexdigest())
-                    shutil.move(f.name, local_path)
-                    url2hash[url_key] = local_path
-            finally:
-                _release_download_cache_lock()
-        else:
-            local_path = f.name
-            if missing_cache:
-                msg = ('File downloaded to temporary location due to problem '
-                       'with cache directory and will not be cached.')
-                warn(CacheMissingWarning(msg, local_path))
-            if conf.delete_temporary_downloads_at_exit:
-                global _tempfilestodel
-                _tempfilestodel.append(local_path)
-    except urllib.error.URLError as e:
-        if hasattr(e, 'reason') and hasattr(e.reason, 'errno') and e.reason.errno == 8:
-            e.reason.strerror = e.reason.strerror + '. requested URL: ' + remote_url
-            e.reason.args = (e.reason.errno, e.reason.strerror)
-        raise e
-    except socket.timeout as e:
-        # this isn't supposed to happen, but occasionally a socket.timeout gets
-        # through.  It's supposed to be caught in `urrlib2` and raised in this
-        # way, but for some reason in mysterious circumstances it doesn't. So
-        # we'll just re-raise it here instead
-        raise urllib.error.URLError(e)
+    if cache:
+        local_path = import_to_cache(url_key, f_name,
+                                     hexdigest=hexdigest,
+                                     remove_original=True)
+    else:
+        local_path = f_name
+        if missing_cache:
+            msg = ('File downloaded to temporary location due to problem '
+                   'with cache directory and will not be cached.')
+            warn(CacheMissingWarning(msg, local_path))
+        if conf.delete_temporary_downloads_at_exit:
+            global _tempfilestodel
+            _tempfilestodel.append(local_path)
 
     return local_path
```

Comment:
It would be nice if maybe download_file could be a thin wrapper around another function that returned an open file object with a name attribute. The other function could be called directly by (or just be?) get_readable_fileobj, and returning an open file object would make it impossible for an ill-timed clear_download_cache to remove the file out from under the user. On the other hand there would be some context-manager jiggery-pokery to get the lock released without closing the file, and really, making the whole interface robust against aggressive use of clear_download_cache would be a major task.